jhub_apps/hub_client/hub_client.py

```diff
@@ -18,21 +18,34 @@
 
 
 class HubClient:
-    def __init__(self, token=None):
-        self.token = token or JUPYTERHUB_API_TOKEN
+    def __init__(self, username=None):
+        self.username = username
+        self.token = JUPYTERHUB_API_TOKEN
         self.jhub_apps_request_id = None
         self._set_request_id()
+        if self.username and JUPYTERHUB_API_TOKEN:
+            self.token = self._create_token_for_user(self.username)
 
     def _set_request_id(self):
         contextvars = structlog.contextvars.get_contextvars()
         self.jhub_apps_request_id = contextvars.get("request_id")
 
-    def _headers(self):
+    def _headers(self, token=None):
         return {
-            "Authorization": f"token {self.token}",
+            "Authorization": f"token {token or self.token}",
             "JHUB_APPS_REQUEST_ID": self.jhub_apps_request_id
         }
 
+    def _create_token_for_user(self, username):
+        logger.info("Creating token for user", username=username)
+        r = requests.post(
+            API_URL + f"/users/{username}/tokens",
+            headers=self._headers(token=JUPYTERHUB_API_TOKEN)
+        )
```

Comment:
The `/tokens` POST endpoint accepts `note`, `expires_in`, `roles` and `scopes`.
- it might be worth setting a `note` to distinguish this token from other tokens when debugging or reviewing the logs (by default it would get "Requested via api" note, maybe "Requested by jhub-apps" could be more useful)
- by default the token has no expiration; this could be fine but possibly sub-optimal from security standpoint; I think it is still an improvement, so maybe expiration handling could be implemented in a follow-up PR, but read on as maybe it is required for the following point
- since the tokens are added to the database, do we need to delete them on shutdown to avoid getting the database size being filled up by tokens? If the shutdown sequence is not guaranteed to run in case of a failure, should we have expiration time anyways so that tokens which are not deleted get automatically invalidated after some time?

Reply:
>it might be worth setting a note to distinguish this token from other tokens when debugging or reviewing the logs (by default it would get "Requested via api" note, maybe "Requested by jhub-apps" could be more useful)

They are currently labelled as: "Requested via api by service japps", can rename that to jhub-apps instead of japps.

>by default the token has no expiration; this could be fine but possibly sub-optimal from security standpoint; I think it is still an improvement, so maybe expiration handling could be implemented in a follow-up PR, but read on as maybe it is required for the following point

>since the tokens are added to the database, do we need to delete them on shutdown to avoid getting the database size being filled up by tokens? If the shutdown sequence is not guaranteed to run in case of a failure, should we have expiration time anyways so that tokens which are not deleted get automatically invalidated after some time?

I thought about tokens overpopulation but usually things are not deleted from db, so deleting them will simply set a flag, so left this to worry about when we have this problem, but this was just my assumption, I just checked and tokens are indeed deleted from database when you revoke them. I agree we should handle this and delete tokens, thanks very valid point indeed!

Reply:
Added a decorator for creation/destruction of user tokens.